// app/views/DataView.js
var View = require("app/views/View");

class DataView extends View {
	constructor(model) {
		super()
		var me = this;

		me._model = model;
		me._model.onChanged.bind(function() {
			me._redraw();
		})
	}
	getHtml() {
		return `
			<div id="${this.uid}" class="dataview">
			</div>
		`;
	}
	_redraw() {
		var acc = [];
		render(this._model.objects, acc);
		this.$element().html(acc.join(""));
	}
}

module.exports = DataView;

var utils = require("app/utils");
var m = require("app/model/ModelClasses");

function render(objects, acc) {
	objects.sort(function(a, b) {
		return (
			((a instanceof m.Apartment) - (b instanceof m.Apartment)) ||
			(a.building - b.building) ||
			((a.section == null) - (b.section == null)) ||
			(a.section - b.section) ||
			(b.floor - a.floor) ||
			(a.number - b.number));
	});

	var lastBuilding = null;
	var lastSection = null;
	var lastFloor = null;

	for (var i = 0, c = objects.length; i < c; i++) {
		var obj = objects[i];
		if (!(obj instanceof m.Apartment))
			continue;

		var building = obj.building;
		var section = obj.section;
		if (building != lastBuilding || section != lastSection) {
			if (lastFloor != null) {
				acc.push("</div>");
				lastFloor = null;
			}
			if (lastBuilding != null) {
				renderBuildingAndSectionInfo(lastBuilding, lastSection, acc);
				acc.push("</div>");
			}

			acc.push("<div class='dataview_section'>");
			lastBuilding = building;
			lastSection = section;
		}

		var floor = obj.floor;
		if (floor != lastFloor) {
			if (lastFloor != null)
				acc.push("</div>");
			acc.push("<div class='dataview_floor'>");
			renderFloorInfo(obj, acc);
			lastFloor = floor;
		}

		renderApartmentInfo(obj, acc);
	}

	acc.push("</div>");
	renderBuildingAndSectionInfo(lastBuilding, lastSection, acc);
	acc.push("</div>");
}

function renderBuildingAndSectionInfo(lastBuilding, lastSection, acc) {
	acc.push(`<div class='dataview_sectionInfo'>Блок ${lastBuilding} Секция ${lastSection == null ? "???" : lastSection}</div>`);
}

function renderFloorInfo(obj, acc) {
	acc.push("<span class='dataview_floorNumber'>");
	acc.push(utils.htmlEncode(obj.floor.toString()));
	acc.push("</span>");
}

function renderApartmentInfo(obj, acc) {
	acc.push(`<div class="dataview_apartment ${obj.originalNumber != null ? "dataview_apartment__withOriginalNumber" : ""}" data-id="${obj.id}">`);
	acc.push(utils.htmlEncode(obj.number.toString()));
	acc.push("</div>");
}
