// app/views/Grid.js
var View = require("app/views/View");

class Grid extends View {
	constructor(model, searchModel, recordSelectionModel) {
		super();
		var me = this;

		me.apartmentPopup = me.addChild(new ApartmentPopup());

		me._model = model;
		me._searchModel = searchModel;
		me._recordSelectionModel = recordSelectionModel;

		me._searchModel.onChanged.bind(function() {
			me._showSearchResults();
		});
		me._recordSelectionModel.onChanged.bind(function() {
			me._showSelectedRecord();
		});
	}
	getHtml() {
		return `
			<div id="${this.uid}" class="grid">
				${this.apartmentPopup.getHtml()}
				<div class="grid_content"></div>
			</div>
		`;
	}
	onInstalled() {
		super.onInstalled(...arguments);
		var me = this;
		me.$element().click(function(ev) {
			var $t = $(ev.target);
			var $e = $t.closest(".grid_apartment");
			if ($e.length) {
				var id = $e[0].id;
				var prefix = me.uid + "_obj-";
				if (id && id.startsWith(prefix)) {
					var id = id.substring(prefix.length);
					me._selectApartment(id, ev);
				}
			}
		});
	}
	redraw() {
		this.apartmentPopup.hide();

		var acc = [];
		render(this.uid, this._model.objects, this._searchModel.getObjectIds(), acc);
		this.$element().children(".grid_content").html(acc.join(""));
	}
	_showSearchResults() {
		var $el = this.$element();
		if (!$el.length)
			return;
		$el.find(".grid_apartment__search").removeClass("grid_apartment__search");
		var ids = this._searchModel.getObjectIds();
		for (var i = 0, c = ids.length; i < c; i++) {
			var id = ids[i];
			$el.find("#" + this.uid + "_obj-" + id).addClass("grid_apartment__search");
		}
	}
	_showSelectedRecord() {
		var $el = this.$element();
		if (!$el.length)
			return;
		$el.find(".grid_apartment__selectedRecord").removeClass("grid_apartment__selectedRecord");
		var record = this._recordSelectionModel.getRecord();
		if (record)
		for (var i = 0, c = record.objects.length; i < c; i++) {
			var obj = record.objects[i];
			$el.find("#" + this.uid + "_obj-" + obj.id).addClass("grid_apartment__selectedRecord");
		}
	}
	_selectApartment(objectId, ev) {
		var object = this._model.getObjectById(objectId);
		if (!object) return;

		this._recordSelectionModel.setRecord(object.record);

		this.apartmentPopup.setObject(object);
		this.apartmentPopup.show(ev);
	}
}

module.exports = Grid;

var utils = require("app/utils");
var m = require("app/model/ModelClasses");
var ApartmentPopup = require("app/views/ApartmentPopup");
var s = require("app/Strings");

function render(gridUid, objects, searchResults, acc) {
	objects.sort(function(a, b) {
		return (
			((a instanceof m.Apartment) - (b instanceof m.Apartment)) ||
			(a.building - b.building) ||
			((a.section == null) - (b.section == null)) ||
			(a.section - b.section) ||
			(b.floor - a.floor) ||
			(a.number - b.number) ||
			((a.record.date != null) - (b.record.date != null)) ||
			(a.record.date - b.record.date));
	});

	var lastBuilding = null;
	var lastSection = null;
	var lastFloor = null;

	for (var i = 0, c = objects.length; i < c; i++) {
		var obj = objects[i];
		if (!(obj instanceof m.Apartment))
			continue;

		var building = obj.building;
		var section = obj.section;
		if (building != lastBuilding || section != lastSection) {
			if (lastFloor != null) {
				acc.push("</div>");
				lastFloor = null;
			}
			if (lastBuilding != null) {
				renderBuildingAndSectionInfo(lastBuilding, lastSection, acc);
				acc.push("</div>");
			}

			acc.push("<div class='grid_section'>");
			lastBuilding = building;
			lastSection = section;
		}

		var floor = obj.floor;
		if (floor != lastFloor) {
			if (lastFloor != null)
				acc.push("</div>");
			acc.push("<div class='grid_floor'>");
			renderFloorInfo(obj, acc);
			lastFloor = floor;
		}

		renderApartmentInfo(gridUid, obj, searchResults, acc);
	}

	acc.push("</div>");
	renderBuildingAndSectionInfo(lastBuilding, lastSection, acc);
	acc.push("</div>");
}

function renderBuildingAndSectionInfo(lastBuilding, lastSection, acc) {
	acc.push(`<div class='grid_sectionInfo'>Корпус ${lastBuilding} Секция ${lastSection == null ? s.Unknown : lastSection}</div>`);
}

function renderFloorInfo(obj, acc) {
	acc.push("<span class='grid_floorNumber'>");
	acc.push(utils.htmlEncode(obj.floor.toString()));
	acc.push("</span>");
}

function renderApartmentInfo(gridUid, obj, searchResults, acc) {
	var cssClasses = ["grid_apartment"];
	if (obj.duplicate)
		cssClasses.push("grid_apartment__duplicate");
	if (obj.originalNumber != null)
		cssClasses.push("grid_apartment__withOriginalNumber");
	if (searchResults.indexOf(obj.id) >= 0)
		cssClasses.push("grid_apartment__search");
	cssClasses = cssClasses.join(" ");

	acc.push(`<div class="${cssClasses}" id="${gridUid + "_obj-" + obj.id}">`);
	acc.push(utils.htmlEncode(obj.number.toString()));
	acc.push("</div>");
}
