// app/views/MainView.js
var View = require("app/views/View");

class MainView extends View
{
	constructor(model, loadController) {
		super();
		this._loadController = loadController;

		this.tabControl = this.addChild(new TabControl());
		this.tabControl.addTab(Tab.Data, "Шахматка");
		this.tabControl.addTab(Tab.Sql, "SQL");
		this.dataView = this.addChild(new DataView(model));
		this.sqlView = this.addChild(new SqlView(model.sqlModel));
	}
	getHtml() {
		return `
			<div id="${this.uid}">
				<ul id="${this.uid + "_messages"}"></ul>
				<button id="${this.uid + "_reload"}">Перезагрузить</button>
				${this.tabControl.getHtml()}
				${this.dataView.getHtml()}
				${this.sqlView.getHtml()}
			</div>
		`;
	}
	onInstalled() {
		super.onInstalled(...arguments);
		var me = this;

		this.$element().find("#" + this.uid + "_reload").click(function() {
			me.load();
		});
	}
	init() {
		this.load();
	}
	load() {
		var $messages = this.$element().find("#" + this.uid + "_messages");
		$messages.empty();
		this._loadController.load().then(null, function(ex) {
			$messages.html("<li>" + utils.htmlEncode(ex.message) + "</li>");
		});
	}
}

var Tab = {
	Data: 1,
	Sql: 2,
};

module.exports = MainView;

var utils = require("app/utils");
var TabControl = require("app/views/TabControl");
var DataView = require("app/views/DataView");
var SqlView = require("app/views/SqlView");
