// app/LoadController.js
class LoadController {
	constructor(model) {
		this._model = model;

		this.participantsRegistryUrl = "data/ЕГРП.csv";
		this.ownersRegistryUrl = "data/Собственники.csv";
		this.apartmentsWithoutSectionUrl = "data/Квартиры без номера секции.csv";
		this.oldApartmentNumbersUrl = "data/Старые номера квартир.csv";

		this._participantsRegistry = null;
		this._ownersRegistry = null;
		this._apartmentsWithoutSection = null;
		this._oldApartmentNumbers = null;

		this._ajaxLoader = new AjaxLoader();
		this._fileLoader = new FileLoader();

		this._opCounter = 0;
		this.onOperationStart = new utils.Delegate();
		this.onOperationEnd = new utils.Delegate();
	}

	// load default data
	init() /* -> Deferred */ {
		var me = this;
		return me._operation(() => $.Deferred().resolve()
			.then(() => me.loadDefaultParticipantsRegistry())
			.then(() => me.loadDefaultOwnersRegistry())
			.then(() => me.loadDefaultApartmentsWithoutSection())
			.then(() => me.loadDefaultOldApartmentNumbers())
			.then(() => me.updateModel()));
	}

	loadDefaultParticipantsRegistry() /* -> Deferred */ {
		var me = this;
		return me._operation(() => me._ajaxLoader.load(this.participantsRegistryUrl)
			.then(function(csv) { me._participantsRegistry = csv; }));
	}
	loadDefaultOwnersRegistry() /* -> Deferred */ {
		var me = this;
		return me._operation(() => me._ajaxLoader.load(this.ownersRegistryUrl)
			.then(function(csv) { me._ownersRegistry = csv; }));
	}
	loadDefaultApartmentsWithoutSection() /* -> Deferred */ {
		var me = this;
		return me._operation(() => me._ajaxLoader.load(this.apartmentsWithoutSectionUrl)
			.then(function(csv) { me._apartmentsWithoutSection = csv; }));
	}
	loadDefaultOldApartmentNumbers() /* -> Deferred */ {
		var me = this;
		return me._operation(() => me._ajaxLoader.load(this.oldApartmentNumbersUrl)
			.then(function(csv) { me._oldApartmentNumbers = csv; }));
	}

	loadApartmentsWithoutSection(file) /* -> Deferred */ {
		var me = this;
		return me._operation(() => me._fileLoader.load(file)
			.then(function(csv) { me._apartmentsWithoutSection = csv; }));
	}
	loadOldApartmentNumbers(file) /* -> Deferred */ {
		var me = this;
		return me._operation(() => me._fileLoader.load(file)
			.then(function(csv) { me._oldApartmentNumbers = csv; }));
	}

	updateModel() /* -> Deferred */ {
		var me = this;
		var model = new Model();

		return me._operation(() => $.Deferred().resolve()
			.then(() => read("данные из ЕГРП", me._participantsRegistry, new ParticipantsRegistryReader(model)))
			.then(() => read("данные из реестра собственникам долей", me._ownersRegistry, new OwnersRegistryReader(model)))
			.then(() => read("данные по квартирам без номера секции", me._apartmentsWithoutSection, new ApartmentsWithoutSectionReader(model)))
			.then(() => read("данные по старым номерам квартир", me._oldApartmentNumbers, new OldApartmentNumbersReader(model)))
			.then(function() {
				try { model.finish(); } catch(ex) { return $.Deferred().reject(ex); }
				me._model.swap(model);
				me._model.changed();
			}));
	}

	_operation(f) {
		var me = this;
		me._operationStart();
		return f().then(
			function() { me._operationEnd(); },
			function(ex) { me._operationEnd(ex); });
	}
	_operationStart() {
		if (this._opCounter++ == 0)
			this.onOperationStart.trigger();
	}
	_operationEnd(ex) {
		this._opCounter = Math.max(1, this._opCounter);
		if (--this._opCounter == 0)
			this.onOperationEnd.trigger(ex);
	}
}

module.exports = LoadController;

var utils = require("app/utils");
var AjaxLoader = require("app/AjaxLoader");
var FileLoader = require("app/FileLoader");
var Model = require("app/model/Model");
var ParticipantsRegistryReader = require("app/model/readers/ParticipantsRegistryReader");
var OwnersRegistryReader = require("app/model/readers/OwnersRegistryReader");
var ApartmentsWithoutSectionReader = require("app/model/readers/ApartmentsWithoutSectionReader");
var OldApartmentNumbersReader = require("app/model/readers/OldApartmentNumbersReader");

function read(dataDescription, data, reader) /* -> Deferred */ {
	var d = $.Deferred();
	if (!data)
		return d.reject(new Error("Отсутствуют " + dataDescription));
	reader.read(data, function(ex) {
		ex ? d.reject(ex) : d.resolve();
	});
	return d;
}
