// app/model/readers/ParticipantsRegistryReader.js
var BaseCsvReader = require("app/model/readers/BaseCsvReader");

class ParticipantsRegistryReader extends BaseCsvReader {
	constructor(model) {
		super({
			skipRows: 1,
		});
		this._model = model;
		this._records = {}; // recordNumber => _Record
	}
	read(csv, callback) {
		var me = this;
		super.read(csv, function(ex) {
			if (!ex)
				for (var key in me._records)
					if (me._records.hasOwnProperty(key))
						me._records[key].finish();
			callback(ex);
		});
	}
	// @override
	_processRecord(record) {
		var recordNumber = record[3];
		if (!recordNumber)
			// не обрабатываем строки без номера записи
			return;

		recordNumber = utils.parseInt(recordNumber);
		if (recordNumber == null)
			throw new Error("Некорректный номер записи: " + record[2]);

		var source = record[1];
		if (!source)
			throw new Error("Отсутвует строка записи");
		var _record;
		var modelRecord;
		if (!(recordNumber in this._records)) {
			_record = this._records[recordNumber] = new _Record(recordNumber);
			modelRecord = _record.modelRecord();
			this._model.addRecord(modelRecord);
		}
		else {
			_record = this._records[recordNumber];
			modelRecord = _record.modelRecord();
		}
		_record.appendSource(source);

		var type = record[4];
		if (!type) {
			// строка без объекта
			for (var i = 5; i < record.length; i++)
				if (record[i])
					throw new Error("Строка, не содержащая тип объекта, не должна содержать информации об объекте");
			return;
		}

		var number = record[5];
		// номер объекта опциональный для нежилых помещений
		if (!number || number == "бн")
			number = null;
		else {
			number = utils.parseInt(number);
			if (number == null)
				throw new Error("Некорректный номер объекта: " + record[5]);
		}

		var building = utils.parseInt(record[6]);
		if (building != 1 && building != 2)
			throw new Error("Некорректный номер корпуса: " + record[6]);

		var area = BaseCsvReader.parseCsvFloat(record[9]);
		if (area == null)
			throw new Error("Некорректное значение площади: " + record[9]);

		if (type == "машиноместо") {
			if (number == null)
				throw new Error("Некорректный номер объекта: " + record[5]);
			this._model.addObject(new m.ParkingPlace(modelRecord, number, building, area));
			return;
		}

		var section = record[8];
		// номер секции опциональный
		if (!section || section == "нет" || section == "?")
			section = null;
		else {
			section = utils.parseInt(section);
			if (section == null)
				throw new Error("Некорректный номер секции: " + record[8]);
		}

		if (type == "неж пом" || type.search(/предприятие/i) >= 0 || type == "магазин" || type == "офис") {
			this._model.addObject(new m.NonResidentialPremise(modelRecord, type, number, building, section, area));
			return;
		}

		var floor = utils.parseInt(record[7]);
		if (floor == null)
			throw new Error("Некорректный номер этажа: " + record[7]);

		if (number == null)
			throw new Error("Некорректный номер объекта: " + record[5]);

		var landingNumber = null;
		this._model.addObject(new m.Apartment(modelRecord, type, number, building, floor, landingNumber, section, area));
	}
}

module.exports = ParticipantsRegistryReader;

var utils = require("app/utils");
var m = require("app/model/ModelClasses.js");

class _Record {
	constructor(recordNumber) {
		this._source = [];
		this._modelRecord = new m.ParticipantsRegistryRecord(recordNumber, null, null, null, null);
	}
	modelRecord() {
		return this._modelRecord;
	}
	appendSource(source) {
		this._source.push(source);
	}
	finish() {
		var source = this._source.join("\n");

		var registryInfo = extractRegistryInfo(source);
		var owner = extractOwner(source);

		this._modelRecord.registryNumber = registryInfo.number;
		this._modelRecord.date = registryInfo.date;
		this._modelRecord.owner = owner;
		this._modelRecord.source = source;
	}
}

function extractRegistryInfo(source) {
	var m = source.match(/№\s*(.*?)\s+от\s+(\d+)\.(\d+)\.(\d+)/);
	if (!m)
		throw new Error("В записи не найдены регистрационный номер и дата регистрации");

	var number = m[1];

	var day = utils.parseInt(m[2]);
	var month = utils.parseInt(m[3]);
	var year = utils.parseInt(m[4]);
	var date = new Date(year, month-1, day);

	return { number, date };
}

function extractOwner(source) /* -> optional string */ {
	var m = source.match(/Участники долевого\n(.*?)(\n|$)/i);
	if (!m)
		throw new Error("В записи не найдены данные об участниках долевого строительства");

	var data = m[1];
	if (!data.match(/общество/i) && !data.match(/юридическое лицо/i)) {
		// физическое лицо
		return data;
	}
	// юридическое лицо
	return null;
}
